cli_tools/common/utils/logging/firelog_logger.go

```diff
@@ -0,0 +1,311 @@
+	logEvent := &ComputeImageToolsLogExtension{
+		ID:           l.ID,
+		CloudBuildID: os.Getenv(daisyutils.BuildIDOSEnv),
+		ToolAction:   l.Action,
+		Status:       statusSuccess,
+		InputParams:  &l.Params,
+		OutputInfo:   getOutputInfo(w, nil),
+	}
+
+	return logEvent, l.sendLogToServer(logEvent)
+}
+
+// logFailure logs a "failure" info to server
+func (l *FirelogLogger) logFailure(err error, w *daisy.Workflow) (*ComputeImageToolsLogExtension, logResult) {
+	logEvent := &ComputeImageToolsLogExtension{
+		ID:           l.ID,
+		CloudBuildID: os.Getenv(daisyutils.BuildIDOSEnv),
+		ToolAction:   l.Action,
+		Status:       statusFailure,
+		InputParams:  &l.Params,
+		OutputInfo:   getOutputInfo(w, err),
+	}
+
+	return logEvent, l.sendLogToServer(logEvent)
+}
+
+func getFailureReason(err error) string {
+	return daisyutils.RemovePrivacyLogTag(err.Error())
+}
+
+func getAnonymizedFailureReason(err error) string {
+	derr := daisy.ToDError(err)
+	if derr == nil {
+		return ""
+	}
+	anonymizedErrs := []string{}
+	for _, m := range derr.AnonymizedErrs() {
+		anonymizedErrs = append(anonymizedErrs, daisyutils.RemovePrivacyLogInfo(m))
+	}
+	return strings.Join(anonymizedErrs, "\n")
+}
+
+func getOutputInfo(w *daisy.Workflow, err error) *OutputInfo {
+	o := OutputInfo{
+		TargetSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(targetSizeGb)),
+		SourceSizeGb: getInt64Value(w.GetSerialConsoleOutputValue(sourceSizeGb)),
+	}
+
+	if err != nil {
+		o.FailureMessage = getFailureReason(err)
+		o.FailureMessageWithoutPrivacyInfo = getAnonymizedFailureReason(err)
+	}
+
+	return &o
+}
+
+func getInt64Value(s string) int64 {
+	i, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (l *FirelogLogger) runWithServerLogging(function func() (*daisy.Workflow, error)) *ComputeImageToolsLogExtension {
+	var event *ComputeImageToolsLogExtension
+
+	// Send log asynchronously. No need to interrupt the main flow when failed to send log, just
+	// keep moving.
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		l.logStart()
+	}()
+
+	if w, err := function(); err != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logFailure(err, w)
+			log.Println(event.OutputInfo.FailureMessage)
+		}()
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logSuccess(w)
+		}()
+	}
+
+	wg.Wait()
+	return event
+}
+
+// RunWithServerLogging runs the function with server logging
+func RunWithServerLogging(action string, params InputParams, function func() (*daisy.Workflow, error)) {
+	l := NewFirelogLogger(action, params)
+	l.runWithServerLogging(function)
+}
+
+func (l *FirelogLogger) sendLogToServer(logEvent *ComputeImageToolsLogExtension) logResult {
+	return l.sendLogToServerWithRetry(logEvent, 3)
+}
+
+func (l *FirelogLogger) sendLogToServerWithRetry(logEvent *ComputeImageToolsLogExtension, maxRetry int) logResult {
```

Comment:
AFAICT, there's no batching--this just sends logs almost in near-realtime. Batching would be prefered, rather than creating a network request for each log. next_request_wait_millis should be more than 5 seconds, which might have the side-effect of having to do a bit of batching.

Reply:
For each command, there are at most 2 calls - one "start" when it's launched and another one "success" or "failure" when it's finished. Batching them is nice to have for perf reason, but I didn't do it because our tool can be killed without any signal. It's mostly executed in "cloudbuild" as a docker, and the cloudbuild itself can be killed. Thus we will lose the chance to log the "start", and lose the history forever.